libutils/fsattrs.c

```diff
@@ -0,0 +1,143 @@
+#include <platform.h>
+#include <fsattrs.h>
+#include <logging.h>
+#include <file_lib.h>
+
+#ifdef HAVE_LINUX_FS_H
+#include <linux/fs.h> /* FS_* constants */
+#endif                /* HAVE_LINUX_FS_H */
+
+
+/**
+ * @brief Get / Set immutable flag for file
+ * @param filename The filename
+ * @param flag The flag
+ * @param get_flag Whether to get (i.e., true) or set (i.e., false) the flag
+ * @return Error code (see FSAttrsResult)
+ * @note Only regular files are currently supported
+ */
+static FSAttrsResult ImmutableFlag(
+    const char *filename, bool *flag, bool get_flag)
+{
+    assert(filename != NULL);
+    assert(flag != NULL);
+
+#if defined(HAVE_LINUX_FS_H)
+
+    /* Since we have to open the file, to use ioctl() we must make sure that
+     * it is in fact a regular file before doing so. E.g., a FIFO file could
+     * cause open(2) to block indefinitely. */
+
+    struct stat sb;
+    if (lstat(filename, &sb) == -1)
+    {
+        return (errno == ENOENT) ? FS_ATTRS_DOES_NOT_EXIST : FS_ATTRS_FAILURE;
+    }
+    if (!S_ISREG(sb.st_mode))
+    {
+        return FS_ATTRS_NOT_SUPPORTED;
+    }
+
+    int fd = safe_open(filename, O_RDONLY);
+    if (fd == -1)
+    {
+        return (errno == ENOENT) ? FS_ATTRS_DOES_NOT_EXIST : FS_ATTRS_FAILURE;
+    }
+
+    int attrs;
+    int ret = ioctl(fd, FS_IOC_GETFLAGS, &attrs);
+    if (ret < 0)
+    {
+        close(fd);
+        return (errno == ENOTTY) ? FS_ATTRS_NOT_SUPPORTED : FS_ATTRS_FAILURE;
+    }
+
+    if (get_flag)
+    {
+        close(fd);
+        *flag = attrs & FS_IMMUTABLE_FL;
+        return FS_ATTRS_SUCCESS;
+    }
+
+    if (*flag)
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Setting immutable bit in inode flags for file '%s'",
+            filename);
+        attrs |= FS_IMMUTABLE_FL;
+    }
+    else
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Clearing immutable bit in inode flags for file '%s'",
+            filename);
+        attrs &= ~FS_IMMUTABLE_FL;
+    }
+
+    ret = ioctl(fd, FS_IOC_SETFLAGS, &attrs);
```

Comment:
Please double-check this works on an FD that's opened with `RDONLY`.

Reply:
It works on my machine :wink: 

```
$ sudo /var/cfengine/bin/cf-agent -Kdf /ntech/polly/immutable.cf | grep "immutable bit"
   debug: Clearing immutable bit in inode flags for file '/tmp/immutable.txt'
 verbose: Temporarily cleared immutable bit for file '/tmp/immutable.txt'
   debug: Setting immutable bit in inode flags for file '/tmp/immutable.txt'
 verbose: Reset immutable bit for file '/tmp/immutable.txt' after temporarily clearing it
```